### PR TITLE
Create a dummy function to prevent NULL miscomparisons, if necessary

### PR DESCRIPTION
**Pull request by ddcc**
Resolves https://github.com/WebAssembly/spec/issues/312

### CODE REVIEW

File: test/local-revision [hunk 1]
```diff
@@ -0,0 +1 @@
+8774
```
Comment (kripken) [inline test/local-revision]: why is this here?

File: src/wasm-linker.cpp [hunk 2]
```diff
@@ -366,10 +369,30 @@ bool hasI64ResultOrParam(FunctionType* ft) {
...
+  for (auto& relocation : out.relocations) {
+    if (relocation->kind == LinkerObject::Relocation::kFunction) {
+      create = true;
+      break;
+    }
+  }
+  if (!create) return;
+  wasm::Builder wasmBuilder(out.wasm);
+  Expression *unreachable = wasmBuilder.makeUnreachable();
+  Function *dummy = wasmBuilder.makeFunction(Name(dummyFunction), std::vector<NameType>(), WasmType::none, std::vector<NameType>(), unreachable);
```
Comment (kripken) [inline src/wasm-linker.cpp]: I think in c++11 you can just do `{}` instead of those `std::vector<NameType>()`s.

File: src/wasm-linker.cpp [hunk 3]
```diff
@@ -366,10 +369,30 @@ bool hasI64ResultOrParam(FunctionType* ft) {
...
+  functionIndexes[dummy->name] = 0;
+  out.wasm.table.names.push_back(dummy->name);
```
Comment (kripken) [inline src/wasm-linker.cpp]: instead of adding to the table manually, you can do `ensureFunctionIndex`. perhaps it should be moved from an `auto` in a method to a proper method on the object.

File: src/wasm-linker.cpp [hunk 4]
```diff
@@ -366,10 +369,30 @@ bool hasI64ResultOrParam(FunctionType* ft) {
   return false;
 }
 
+void Linker::makeDummyFunction() {
+  bool create = false;
+  // Check if there are indirectly called functions
```
Comment (kripken) [inline src/wasm-linker.cpp]: is it really impossible to get a comparison to a function pointer without this? couldn't one compare `(void*)17` to `nullptr` for example? I guess that's a weird thing we don't care about much ;) but i wonder if we shouldn't be totally general.

File: test/local-revision (hunk 1 above)
Comment (ddcc) [inline test/local-revision]: Oops, that's what #657 is for :)

**Comment by ddcc** [comment]
There's a bit of a trade-off on the dummy function. Is it something we always want to create, even if no functions are address taken? The only case that a function indexes is compared against is if either lhs/rhs is a function.

File: src/wasm-linker.cpp (hunk 4 above)
Comment (dschuff) [inline src/wasm-linker.cpp]: This comment should say 'address-taken' functions, since you can compare pointers without calling anything :)

**Comment by dschuff** [comment]
I think we shouldn't create the function if there are no address-taken functions (because then we can have no table at all). Casting a random number to a function pointer is UB. Practically speaking, compilers will usually let you do it, especially if you just call them (e.g. when you have things mapped into memory like VDSOs or NaCl's syscall trampolines). However if you just casted a number like 17 and compared it to nullptr, you still wouldn't have anything be address-taken, you'd just get integer arithmetic.

File: src/wasm-linker.cpp [hunk 5]
```diff
@@ -387,10 +389,29 @@ bool hasI64ResultOrParam(FunctionType* ft) {
...
+void Linker::makeDummyFunction() {
+  bool create = false;
+  // Check if there are address-taken functions
+  for (auto& relocation : out.relocations) {
+    if (relocation->kind == LinkerObject::Relocation::kFunction) {
+      create = true;
+      break;
+    }
+  }
+  if (!create || out.wasm.table.names.size()) return;
```
Comment (dschuff) [inline src/wasm-linker.cpp]: This check means that if there's anything in the table, we bail, which I guess is for the case where the indexes are pre-assigned. Otherwise if that happens, we are basically already in a screwed-up state, right? Can we put an assert before we run the pre-assignment that the table is empty?

File: src/wasm-linker.h [hunk 6]
```diff
@@ -252,6 +252,9 @@ class Linker {
   // Returns false if an error occurred.
   bool linkArchive(Archive& archive);
 
+  // Name of the dummy function.
+  static constexpr const char* dummyFunction = "__wasm_dummy";
```
Comment (dschuff) [inline src/wasm-linker.h]: let's make this `__wasm_nullptr` or something a little bit more descriptive.

**Comment by dschuff** [comment]
2 more comments, otherwise LGTM

File: src/wasm-linker.cpp (hunk 5 above)
Comment (ddcc) [inline src/wasm-linker.cpp]: This function should always be called before the functions with pre-assigned indices are handled, since the pre-assigned functions will  start at index 1, so the two should be mutually compatible.  But the intention is to check just in case the table is non-empty when this function is called,  so I will convert it to an assert.

**Comment by ddcc** [comment]
Should be good now.

File: src/wasm-linker.cpp [hunk 7]
```diff
@@ -387,10 +389,30 @@ bool hasI64ResultOrParam(FunctionType* ft) {
   return false;
 }
 
+void Linker::makeDummyFunction() {
+  assert(!out.wasm.table.names.size());
```
Comment (dschuff) [inline src/wasm-linker.cpp]: Oh, and let's make this `assert(out.wasm.table.names.empty())`, that's more readable.
